cpp/dolfinx/geometry/BoundingBoxTree.cpp

```diff
@@ -309,29 +298,18 @@ BoundingBoxTree::BoundingBoxTree(const mesh::Mesh& mesh, int tdim) : _tdim(tdim)
   if (mpi_size > 1)
   {
     // Send root node coordinates to all processes
-    std::vector<double> send_bbox(bbox_coordinates.end() - 6,
-                                  bbox_coordinates.end());
-    std::vector<double> recv_bbox(send_bbox.size() * mpi_size);
-    MPI_Allgather(send_bbox.data(), send_bbox.size(), MPI_DOUBLE,
-                  recv_bbox.data(), send_bbox.size(), MPI_DOUBLE, comm);
-
-    std::vector<int> global_leaves(mpi_size);
-    std::iota(global_leaves.begin(), global_leaves.end(), 0);
-    global_tree.reset(new BoundingBoxTree(recv_bbox, global_leaves.begin(),
-                                          global_leaves.end()));
+    const auto send_bbox = _bbox_coordinates.bottomRows(2);
+    Eigen::Array<double, Eigen::Dynamic, 3, Eigen::RowMajor> recv_bbox(
+        mpi_size * 2, 3);
+    MPI_Allgather(send_bbox.data(), 6, MPI_DOUBLE, recv_bbox.data(), 6,
+                  MPI_DOUBLE, comm);
+
+    auto [global_bboxes, global_coords] = build_from_leaf(recv_bbox);
+    global_tree.reset(new BoundingBoxTree(global_bboxes, global_coords));
```

Comment:
Use `_global_free = std::make_unique<BoundingBoxTree>(global_bboxes, global_coords);`

Reply:
Actually, that won't work, because the constructor is private...